lib/uplink/example_test.go

```diff
@@ -0,0 +1,186 @@
+		defer func() { err = errs.Combine(err, upl.Close()) }()
+
+		// It is temporarily required to set the encryption key in project options.
+		// This requirement will be removed in the future.
+		opts := uplink.ProjectOptions{}
+		opts.Volatile.EncryptionKey = encryptionKey
+
+		// Open up the Project we will be working with
+		proj, err := upl.OpenProject(ctx, satelliteAddr, parsedAPIKey, &opts)
+		if err != nil {
+			return fmt.Errorf("could not open project: %+v", err)
+		}
+		defer func() { err = errs.Combine(err, proj.Close()) }()
+
+		// Set the bucket's redundancy scheme to the one of the test planet's uplink.
+		// This is only necessary for test planet, but not for real use of libuplink.
+		bucketCfg := uplink.BucketConfig{}
+		bucketCfg.Volatile.RedundancyScheme = redundancy
+
+		// Create the desired Bucket within the Project
+		_, err = proj.CreateBucket(ctx, testBucket, &bucketCfg)
+		if err != nil {
+			return fmt.Errorf("could not create bucket: %+v", err)
+		}
+
+		// Open up the desired Bucket within the Project
+		bucket, err := proj.OpenBucket(ctx, testBucket, &uplink.EncryptionAccess{Key: *encryptionKey})
+		if err != nil {
+			return fmt.Errorf("could not open bucket: %+v", err)
+		}
+		defer func() { err = errs.Combine(err, bucket.Close()) }()
+
+		testData, err := ioutil.ReadAll(io.LimitReader(rand.Reader, testDataSize.Int64()))
+		if err != nil {
+			return fmt.Errorf("could not generate test data: %+v", err)
+		}
+
+		// Upload our Object to the specified path
+		buf := bytes.NewBuffer(testData)
+		err = bucket.UploadObject(ctx, testUploadPath, buf, nil)
+		if err != nil {
+			return fmt.Errorf("could not upload: %+v", err)
+		}
+
+		// Initiate a download of the same object again
+		readBack, err := bucket.OpenObject(ctx, testUploadPath)
+		if err != nil {
+			return fmt.Errorf("could not open object: %+v", err)
+		}
+		defer func() { err = errs.Combine(err, readBack.Close()) }()
+
+		// We want the whole thing, so range from 0 to -1
+		strm, err := readBack.DownloadRange(ctx, 0, -1)
+		if err != nil {
+			return fmt.Errorf("could not initiate download: %+v", err)
+		}
+		defer func() { err = errs.Combine(err, strm.Close()) }()
+
+		// Read everything from the stream
+		receivedContents, err := ioutil.ReadAll(strm)
+		if err != nil {
+			return fmt.Errorf("could not read object: %+v", err)
+		}
+
+		if !bytes.Equal(receivedContents, testData) {
+			return fmt.Errorf("got different object back: %q != %q", testData, receivedContents)
+		}
+
+		// Delete the Object
+		err = bucket.DeleteObject(ctx, testUploadPath)
+		if err != nil {
+			return fmt.Errorf("could not delete object: %+v", err)
+		}
+
+		// Delete the Bucket
+		err = proj.DeleteBucket(ctx, testBucket)
+		if err != nil {
+			return fmt.Errorf("could not delete bucket: %+v", err)
+		}
+
+		return nil
+	})
+
+	fmt.Println("success!")
+
+	// Output:
+	// success!
+}
+
+func setupTestEnv(test func(context.Context, string, string, storj.RedundancyScheme) error) {
```

Comment:
rather than use callbacks here return the planet, satellite address, api key and redundancy